# Projeto SQL — Análise de Banco de Dados de Livros

## Introdução

A pandemia do coronavírus provocou mudanças significativas nos hábitos das pessoas, aumentando o tempo dedicado a atividades realizadas em casa, como a leitura. Esse cenário impulsionou o crescimento do mercado de livros digitais e despertou o interesse de diversas startups em desenvolver novos aplicativos voltados para leitores.

Neste contexto, foi disponibilizado um banco de dados de um serviço concorrente contendo informações sobre livros, autores, editoras, classificações e avaliações realizadas pelos usuários. A análise desses dados permitirá compreender melhor as características do catálogo e o comportamento dos leitores.

O objetivo deste projeto é explorar o banco de dados por meio de consultas SQL e obter informações relevantes que possam contribuir para a definição de uma proposta de valor para um novo produto voltado ao mercado de leitura digital. Todas as análises serão realizadas utilizando SQL, enquanto o Pandas será utilizado apenas para exibição e armazenamento dos resultados no Jupyter Notebook.


## Objetivos do Estudo

Os principais objetivos deste estudo são:

- Explorar a estrutura e o conteúdo das tabelas disponíveis no banco de dados;
- Compreender as relações entre livros, autores, editoras, classificações e avaliações;
- Responder às questões de negócio propostas por meio de consultas SQL;
- Identificar padrões relevantes relacionados às publicações e ao engajamento dos usuários;
- Apresentar conclusões baseadas nos resultados obtidos.

In [ ]:
# Database connection
# Credentials are loaded from environment variables to avoid exposing private data.

import os
import pandas as pd
from sqlalchemy import create_engine

db_config = {
    "user": os.getenv("DB_USER"),
    "pwd": os.getenv("DB_PASSWORD"),
    "host": os.getenv("DB_HOST"),
    "port": os.getenv("DB_PORT"),
    "db": os.getenv("DB_NAME")
}

connection_string = "postgresql://{}:{}@{}:{}/{}".format(
    db_config["user"],
    db_config["pwd"],
    db_config["host"],
    db_config["port"],
    db_config["db"]
)

engine = create_engine(connection_string, connect_args={"sslmode": "require"})

## Exploração Inicial das Tabelas

Antes de realizar as análises propostas, será feita uma exploração inicial das tabelas disponíveis no banco de dados. Essa etapa permite compreender a estrutura dos dados, identificar as variáveis disponíveis e entender como as tabelas se relacionam.

### Tabela books

A tabela `books` contém informações sobre os livros disponíveis na plataforma, incluindo título, número de páginas, data de publicação, autor e editora.

In [8]:
query = '''
SELECT *
FROM books
LIMIT 5
'''

df = pd.read_sql(query, con=engine)
df

,book_id,author_id,title,num_pages,publication_date,publisher_id
0,1,546,'Salem's Lot,594,2005-11-01,93
1,2,465,1 000 Places to See Before You Die,992,2003-05-22,336
2,3,407,13 Little Blue Envelopes (Little Blue Envelope...,322,2010-12-21,135
3,4,82,1491: New Revelations of the Americas Before C...,541,2006-10-10,309
4,5,125,1776,386,2006-07-04,268


### Tabela authors

A tabela `authors` contém a identificação e o nome dos autores presentes na base de dados.

In [10]:
query = '''
SELECT *
FROM authors
LIMIT 5
'''

pd.read_sql(query, con=engine)

,author_id,author
0,1,A.S. Byatt
1,2,Aesop/Laura Harris/Laura Gibbs
2,3,Agatha Christie
3,4,Alan Brennert
4,5,Alan Moore/David Lloyd


### Tabela publishers

A tabela `publishers` contém informações sobre as editoras responsáveis pela publicação dos livros.

In [11]:
query = '''
SELECT *
FROM publishers
LIMIT 5
'''

pd.read_sql(query, con=engine)

,publisher_id,publisher
0,1,Ace
1,2,Ace Book
2,3,Ace Books
3,4,Ace Hardcover
4,5,Addison Wesley Publishing Company


### Tabela ratings

A tabela `ratings` registra as classificações atribuídas pelos usuários aos livros.

In [25]:
query = '''
SELECT *
FROM ratings
LIMIT 5
'''

pd.read_sql(query, con=engine)

,rating_id,book_id,username,rating
0,1,1,ryanfranco,4
1,2,1,grantpatricia,2
2,3,1,brandtandrea,5
3,4,2,lorichen,3
4,5,2,mariokeller,2


### Tabela reviews

A tabela `reviews` contém as avaliações textuais escritas pelos usuários sobre os livros.

In [12]:
query = '''
SELECT *
FROM reviews
LIMIT 5
'''

pd.read_sql(query, con=engine)

,review_id,book_id,username,text
0,1,1,brandtandrea,Mention society tell send professor analysis. ...
1,2,1,ryanfranco,Foot glass pretty audience hit themselves. Amo...
2,3,2,lorichen,Listen treat keep worry. Miss husband tax but ...
3,4,3,johnsonamanda,Finally month interesting blue could nature cu...
4,5,3,scotttamara,Nation purpose heavy give wait song will. List...


## Relação entre as Tabelas

Com base na estrutura observada, verifica-se que a tabela `books` é a tabela central do banco de dados.

As relações identificadas são:

- `books.author_id` → `authors.author_id`
- `books.publisher_id` → `publishers.publisher_id`
- `ratings.book_id` → `books.book_id`
- `reviews.book_id` → `books.book_id`

Dessa forma, cada livro pode ser associado ao seu autor, à sua editora, às classificações recebidas e às avaliações escritas pelos usuários.

In [15]:
query = '''
SELECT COUNT(*) AS total_books
FROM books
'''

pd.read_sql(query, con=engine)


,total_books
0,1000


In [16]:
query = '''
SELECT COUNT(*) AS total_authors
FROM authors
'''

pd.read_sql(query, con=engine)

,total_authors
0,636


In [17]:
query = '''
SELECT COUNT(*) AS total_publishers
FROM publishers
'''

pd.read_sql(query, con=engine)

,total_publishers
0,340


In [18]:
query = '''
SELECT COUNT(*) AS total_ratings
FROM ratings
'''

pd.read_sql(query, con=engine)

,total_ratings
0,6456


In [19]:
query = '''
SELECT COUNT(*) AS total_reviews
FROM reviews
'''

pd.read_sql(query, con=engine)

,total_reviews
0,2793


In [28]:
#Relacao entre tabelas 

query = '''
SELECT COUNT(*) AS total_books
FROM books
'''
pd.read_sql(query, engine)

,total_books
0,1000


## Diagnóstico Inicial dos Dados

A base de dados contém 1.000 livros, além de informações relacionadas a autores, editoras, classificações e avaliações dos usuários. A presença dessas tabelas permite analisar tanto as características do catálogo quanto o comportamento dos leitores na plataforma.

## Resolução das Tarefas Propostas

### Tarefa 1: Número de livros publicados após 1º de janeiro de 2000

In [21]:
query = '''
SELECT COUNT(*) AS total_books
FROM books
WHERE publication_date > '2000-01-01'
'''

result = pd.read_sql(query, con=engine)
result

,total_books
0,819


### Conclusão

A consulta identificou 819 livros publicados após 1º de janeiro de 2000.

Considerando que a base de dados contém 1.000 livros no total, observa-se que aproximadamente 81,9% do catálogo é composto por obras publicadas a partir dos anos 2000. Em contrapartida, apenas 181 livros foram publicados antes desse período.

Esses resultados indicam que a plataforma possui uma forte concentração de títulos contemporâneos, sugerindo um catálogo voltado principalmente para publicações mais recentes e alinhadas às tendências atuais do mercado editorial.

### Tarefa 2: Número de avaliações e classificação média para cada livro

O objetivo desta análise é identificar o nível de engajamento dos usuários com cada livro por meio da quantidade de avaliações recebidas, além de calcular a classificação média atribuída pelos leitores.

In [22]:
query = '''
SELECT
    b.book_id,
    b.title,
    COUNT(DISTINCT rv.review_id) AS total_reviews,
    ROUND(AVG(rt.rating), 2) AS avg_rating
FROM books AS b
LEFT JOIN reviews AS rv
    ON b.book_id = rv.book_id
LEFT JOIN ratings AS rt
    ON b.book_id = rt.book_id
GROUP BY
    b.book_id,
    b.title
ORDER BY total_reviews DESC
'''

result = pd.read_sql(query, con=engine)
result.head(10)

,book_id,title,total_reviews,avg_rating
0,948,Twilight (Twilight #1),7,3.66
1,963,Water for Elephants,6,3.98
2,734,The Glass Castle,6,4.21
3,302,Harry Potter and the Prisoner of Azkaban (Harr...,6,4.41
4,695,The Curious Incident of the Dog in the Night-Time,6,4.08
5,696,The Da Vinci Code (Robert Langdon #2),6,3.83
6,627,The Alchemist,6,3.79
7,750,The Hobbit or There and Back Again,6,4.13
8,656,The Book Thief,6,4.26
9,779,The Lightning Thief (Percy Jackson and the Oly...,6,4.08


### Observação dos Resultados

Entre os livros analisados, o maior número de resenhas identificado foi de 7, correspondente ao livro *Twilight (Twilight #1)*. Os demais títulos mais comentados apresentaram 6 resenhas cada, indicando uma diferença relativamente pequena entre os livros com maior engajamento.

Em relação às classificações médias, o livro *Harry Potter and the Prisoner of Azkaban* apresentou a maior avaliação entre os dez primeiros resultados (4,41), seguido por *The Book Thief* (4,26) e *The Glass Castle* (4,21).

### Conclusão

Os resultados sugerem que o engajamento dos leitores está distribuído de forma relativamente equilibrada entre os livros mais populares, sem a existência de um único título concentrando grande parte das resenhas.

Além disso, observa-se que os livros mais comentados apresentam avaliações médias elevadas, geralmente superiores a 4 pontos. Isso pode indicar uma relação positiva entre satisfação dos leitores e participação por meio de avaliações e resenhas, informação relevante para a identificação de obras com maior potencial de destaque dentro da plataforma.

### Tarefa 3: Identificar a editora que publicou o maior número de livros com mais de 50 páginas

Nesta etapa, busca-se identificar a editora responsável pela publicação do maior número de livros com mais de 50 páginas. O critério de páginas foi utilizado para excluir brochuras e publicações muito curtas, concentrando a análise em livros completos.

In [24]:
query = '''
SELECT
    p.publisher,
    COUNT(b.book_id) AS total_books
FROM books AS b
INNER JOIN publishers AS p
    ON b.publisher_id = p.publisher_id
WHERE b.num_pages > 50
GROUP BY p.publisher
ORDER BY total_books DESC
LIMIT 5
'''

pd.read_sql(query, con=engine)

,publisher,total_books
0,Penguin Books,42
1,Vintage,31
2,Grand Central Publishing,25
3,Penguin Classics,24
4,Bantam,19


### Observação dos Resultados

Entre as editoras analisadas, a Penguin Books apresentou o maior número de livros com mais de 50 páginas, totalizando 42 publicações. A segunda colocada foi a Vintage, com 31 livros, resultando em uma diferença de 11 títulos entre as duas editoras.

Esse resultado indica uma liderança relativamente clara da Penguin Books dentro do catálogo analisado.

### Conclusão

A Penguin Books foi identificada como a editora com maior presença entre os livros completos da base de dados. A vantagem de 11 títulos em relação à segunda colocada sugere uma participação relevante da editora no catálogo disponível, contribuindo significativamente para a oferta de livros na plataforma.

### Tarefa 4: Identificar o autor com a maior classificação média

Nesta etapa, busca-se identificar o autor com a maior classificação média de seus livros.

Para garantir maior confiabilidade na análise, serão considerados apenas livros que receberam pelo menos 50 classificações dos usuários. Esse critério foi adotado para reduzir o impacto de livros com poucas avaliações, que poderiam distorcer os resultados e gerar conclusões menos representativas.

In [26]:
query = '''
WITH books_50_ratings AS (
    SELECT
        book_id
    FROM ratings
    GROUP BY book_id
    HAVING COUNT(rating_id) >= 50
)

SELECT
    a.author,
    ROUND(AVG(r.rating), 2) AS avg_rating
FROM books b
INNER JOIN authors a
    ON b.author_id = a.author_id
INNER JOIN ratings r
    ON b.book_id = r.book_id
WHERE b.book_id IN (
    SELECT book_id
    FROM books_50_ratings
)
GROUP BY a.author
ORDER BY avg_rating DESC
LIMIT 10
'''

result = pd.read_sql(query, con=engine)
result

,author,avg_rating
0,J.K. Rowling/Mary GrandPré,4.29
1,Markus Zusak/Cao Xuân Việt Khương,4.26
2,J.R.R. Tolkien,4.25
3,Louisa May Alcott,4.19
4,Rick Riordan,4.08
5,William Golding,3.90
6,J.D. Salinger,3.83
7,William Shakespeare/Paul Werstine/Barbara A. M...,3.79
8,Paulo Coelho/Alan R. Clarke/Özdemir İnce,3.79
9,Lois Lowry,3.75


### Observação dos Resultados

Considerando apenas livros com pelo menos 50 classificações, o autor com a maior média de avaliação foi **J.K. Rowling/Mary GrandPré**, com nota média de **4,29**.

As posições seguintes foram ocupadas por **Markus Zusak/Cao Xuân Việt Khương** (4,26) e **J.R.R. Tolkien** (4,25). Observa-se que a diferença entre os três primeiros colocados é inferior a 0,05 pontos, indicando um nível de aprovação muito semelhante entre esses autores.

Além disso, todos os autores presentes no Top 10 apresentaram médias superiores a **3,75 pontos**, enquanto os cinco primeiros colocados superaram a marca de **4,0 pontos**. Esses resultados sugerem que os livros com maior volume de avaliações tendem a receber classificações positivas dos leitores.


### Conclusão

A análise identificou J.K. Rowling/Mary GrandPré como os autores com a maior classificação média entre os livros que receberam pelo menos 50 avaliações.

A adoção de um número mínimo de classificações aumentou a confiabilidade da análise, reduzindo possíveis distorções causadas por livros com poucas avaliações. Os resultados também sugerem que autores amplamente reconhecidos pelo público tendem a manter elevados níveis de satisfação entre os leitores, o que pode representar uma oportunidade para recomendações e estratégias de destaque dentro da plataforma.

### Tarefa 5: Número médio de avaliações entre usuários que classificaram mais de 50 livros

Nesta etapa, busca-se analisar o comportamento dos usuários mais ativos da plataforma. Para isso, serão considerados apenas os usuários que classificaram mais de 50 livros.

O objetivo é calcular o número médio de avaliações textuais (reviews) produzidas por esse grupo de usuários, permitindo compreender melhor seu nível de engajamento além das classificações numéricas.

In [27]:
query = '''
WITH active_users AS (
    SELECT
        username
    FROM ratings
    GROUP BY username
    HAVING COUNT(book_id) > 50
)

SELECT
    ROUND(AVG(review_count), 2) AS avg_reviews
FROM (
    SELECT
        username,
        COUNT(review_id) AS review_count
    FROM reviews
    WHERE username IN (
        SELECT username
        FROM active_users
    )
    GROUP BY username
) t
'''

result = pd.read_sql(query, con=engine)
result

,avg_reviews
0,24.33


### Observação dos Resultados

A análise mostrou que os usuários que classificaram mais de 50 livros escreveram, em média, **24,33 avaliações textuais**.

Considerando que todos os usuários analisados atribuíram classificações a mais de 50 livros, observa-se que o número médio de resenhas é significativamente menor que o número de classificações realizadas. Isso indica que os usuários tendem a avaliar os livros com notas com maior frequência do que escrevem comentários detalhados sobre suas experiências de leitura.


### Conclusão

Os resultados sugerem que a participação dos usuários mais ativos ocorre principalmente por meio de classificações numéricas, enquanto a produção de avaliações textuais exige um nível maior de envolvimento.

Para a plataforma, esse comportamento indica uma oportunidade de desenvolver mecanismos que incentivem a criação de resenhas, como sistemas de recompensas, recomendações personalizadas ou recursos de interação social. O aumento do número de avaliações textuais pode enriquecer a experiência dos leitores e fornecer informações mais detalhadas sobre os livros disponíveis no catálogo.


# Conclusões Gerais

O presente estudo permitiu analisar informações sobre livros, autores, editoras, classificações e avaliações dos usuários presentes na base de dados.

Os resultados mostraram que **819 dos 1.000 livros da base (81,9%) foram publicados após o ano 2000**, evidenciando uma predominância de obras mais recentes no catálogo. Também foi possível observar que os livros mais comentados apresentam avaliações médias elevadas, com destaque para títulos populares amplamente reconhecidos pelos leitores.

Entre as editoras, a **Penguin Books** destacou-se por possuir o maior número de livros com mais de 50 páginas, totalizando **42 publicações**, superando a segunda colocada, **Vintage**, por uma diferença de 11 títulos.

Em relação aos autores, **J.K. Rowling/Mary GrandPré** apresentou a maior classificação média entre os autores com livros que receberam pelo menos 50 avaliações, alcançando uma nota média de **4,29**. Além disso, todos os autores presentes no Top 10 obtiveram médias superiores a **3,75 pontos**, indicando uma avaliação geralmente positiva por parte dos leitores.

Por fim, verificou-se que os usuários mais ativos da plataforma, definidos como aqueles que classificaram mais de 50 livros, escreveram em média **24,33 avaliações textuais**. Esse resultado sugere que os usuários tendem a utilizar classificações numéricas com maior frequência do que produzir resenhas detalhadas.

De modo geral, os resultados contribuíram para uma melhor compreensão das características do catálogo e do comportamento dos leitores, fornecendo informações relevantes para o desenvolvimento de novos produtos voltados ao mercado de leitura digital.


## Materiais do Projeto

**Notebook de análise:** disponível neste documento.

**Apresentação em PDF:**
https://drive.google.com/file/d/1mStFk1xa-7n-SJmtmIryDaqKDlSdh0a5/view?usp=sharing

A apresentação resume os principais resultados, insights e conclusões obtidos durante a análise do banco de dados.
